README.md
# node-imagickal

node wrapper for ImageMagick commands

[![Build Status](https://secure.travis-ci.org/martinj/node-imagickal.png)](http://travis-ci.org/martinj/node-imagickal)

## Installation

	npm install imagickal

## Examples

```javascript
	var im = require('imagickal');

	//get image dimensions
	im.dimensions('image.jpg').then(function (dim) {
		console.log(dim.width);
		console.log(dim.height);
	});

	//get image dimensions and type
	im.identify('image.jpg').then(function (data) {
		console.log(data);
	});

	//Add true as second argument on identify to check if the image is corrupt.
	im.identify('image.jpg', true).then(function (data) {
		console.log(data);
	});

	//Don't like using promises, use regular node style callbacks for all functions that returns promises.
	im.identify('image.jpg', function (err, data) {
		console.log(data);
	});

	//Using streams instead of filepath
	im.identify(fs.createReadStream('image.jpg'), function (err, data) {
		console.log(data);
	});

	im.transform(fs.createReadStream('image.jpg'), fs.createWriteStream('out.jpg', {encoding: 'binary'}), {strip: true}).then(function () {
		console.log('Done')
	});


	//transform image with action object,
	//actions is applied in the same order as they are recevied
	var actions =  {
		resize: { width: 100 },
		crop: { width: 10, height: 10, x: 10, y: 10 },
		quality: 90,
		strip: true
	};

	im.transform('src.jpg', 'dst.jpg', actions).then(function () {
		console.log('Done')
	});

	//transform image with command object
	im.commands()
		.resize({ width: 100 })
		.crop({ width: 10, height: 10, x: 10, y: 10 })
		.quality(90)
		.strip()
		.exec('src.jpg', 'dst.jpg').then(function () {
			console.log('done');
		});

	//specify executable path
	im.commands({ executable: 'MAGICK_MEMORY_LIMIT=256MB /tmp/convert' }).exec('src.jpg', 'dst.png');

	//or set global for im.
	im.setDefaults({ executable: 'MAGICK_MEMORY_LIMIT=256MB /tmp/convert' });
```

## Available image actions / commands

- rotate Rotate image, properties { angle: 0, x: 0, y: 0 }
- crop - Crop image, properties { width: 2560, height: 1013, x: 0, y: 0 }
- extent - Directly adjust image size, properties { width: 2560, height: 1013 }
- resize - Resize image, properties { width: 470, height: 186, flag: '>' }, available flags <,>,!,^ for more info read about them in the imagemagick [resize docs](http://www.imagemagick.org/Usage/resize/#noaspect)
- sharpen - Sharpen image, properties { mode: 'variable' }, available modes are: light, moderate, strong, extreme, off.
- strip - Strip image of all profiles and comments.
- quality - Compression quality, defaults to 85.

## Run Tests

	npm test

## Enable debugging output

Set DEBUG environment variable to imagickal

	DEBUG=imagickal node ./app.js


CHANGELOG.md
### 5.0.1

* [[f771e46](../../commit/f771e46)] Quotes to paths added

### 5.0.0

* [[d1baf22](../../commit/d1baf22)] Add support for custom output format on `.identify()`. Function signature has been changed.

### 4.2.0

* [[2d0237d](../../commit/2d0237d)] Make DPI-test more portable
* [[8da0ab3](../../commit/8da0ab3)] Support variable density for rendering graphics, add SVG-tests

### 4.1.0

* [[fd820fd](../../commit/fd820fd)] Add support for specifying output format

### 4.0.0

* [[276afc2](../../commit/276afc2)] Variable sharpening will be selected based on the lowest of width or height instead of both. This also fixes a problem with using streams.

### 3.0.0

* [[afb841f](../../commit/afb841f)] Node 8 and support for using streams

### 2.0.0

Now uses bluebird@3.x which may contain breaking changes.

* [adbde5e](../commit/adbde5e) - Change linting.
* [e32356b](../commit/e32356b) - Update travis to test on node v4 instead of iojs.
* [563c3b6](../commit/563c3b6) - Remove use of deprecated bluebird functions.
* [46c25af](../commit/46c25af) - Fix linting and broken test.
* [5ae436a](../commit/5ae436a) - chore(package): update dependencies


test/helper.js
'use strict';
const readChunk = require('read-chunk');
const fileType = require('file-type');

function is(file, extension) {
	const buffer = readChunk.sync(file, 0, fileType.minimumBytes);
	return fileType(buffer).ext === extension;
}

exports.isJPG = function (file) {
	return is(file, 'jpg');
};

exports.isPNG = function (file) {
	return is(file, 'png');
};


test/commands.test.js
'use strict';
const fs = require('fs');
const stream = require('stream');
const Readable = stream.Readable;
const Writeable = stream.Writable;
const ImageMagickCommands = require('../lib/commands');
const { isJPG, isPNG } = require('./helper');

const imageFile = __dirname + '/fixtures/small.jpg';

require('should');

describe('ImageMagickCommands', () => {
	it('should be able to configure custom executable', () => {
		const cmds = new ImageMagickCommands({ executable: 'MAGICK_MEMORY_LIMIT=256MB /tmp/convert' });
		const cmd = cmds.get('src.jpg', 'dst.jpg');
		cmd.should.equal('MAGICK_MEMORY_LIMIT=256MB /tmp/convert "src.jpg" "dst.jpg"');
	});

	it('should apply commands in the same order as its called', () => {
		const cmds = new ImageMagickCommands();
		const cmd = cmds.strip()
			.quality(5)
			.crop({ width: 1, height: 2, x: 3, y: 4 })
			.get('src.jpg', 'dst.jpg');

		cmd.should.equal('convert "src.jpg" -strip -quality 5 -crop 1x2+3+4 "dst.jpg"');
	});

	describe('#rotate', () => {
		it('should not apply when angle is 0', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.rotate({ angle: 0, x: 1, y: 2 }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should set background if supplied', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.rotate({ angle: 1, x: 1, y: 2, bgColor: 'blue' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -background blue -virtual-pixel background -distort ScaleRotateTranslate \'1,2 1\' "dst.jpg"');
		});
	});

	describe('#resize', () => {
		it('should not apply when if width or height is missing', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.resize({ flag: '!' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should quote some flags', () => {
			let cmds = new ImageMagickCommands();
			let cmd = cmds.resize({ width: 10, flag: '>' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -filter Catrom -resize 10x\\> "dst.jpg"');

			cmds = new ImageMagickCommands();
			cmd = cmds.resize({ width: 10, flag: '^' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -filter Catrom -resize 10x^ "dst.jpg"');

		});

		it('should not set flag if invalid', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.resize({ width: 10, flag: 'f' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -filter Catrom -resize 10x "dst.jpg"');
		});
	});

	describe('#gravity', () => {
		it('should set gravity option', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.gravity('Center').get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -gravity Center "dst.jpg"');
		});

		it('should not set gravity option if invalid', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.gravity('Invalid').get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});
	});

	describe('#sharpen', () => {
		it('should be ignored if mode is unknown', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.sharpen({ mode: 'foobar' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should be ignored when mode is off', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.sharpen({ mode: 'off' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should be ignored when mode is falsy', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.sharpen({ mode: 0 }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should be ignored when mode is variable and no width or height is specified', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.sharpen({ mode: 'variable' }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should be choose preset when mode is variable', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.sharpen({ mode: 'variable', width: 250, height: 250 }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -unsharp 0.65x0.65+1.1+0.05 "dst.jpg"');
		});
	});

	describe('#crop', () => {
		it('should handle positive x,y values', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.crop({ width: 100, height: 250, x: 10, y: 0 }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -crop 100x250+10+0 "dst.jpg"');
		});

		it('should handle negative x,y values', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.crop({ width: 100, height: 250, x: -10, y: -12 }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -crop 100x250-10-12 "dst.jpg"');
		});
	});

	describe('#density', () => {
		it('should support variable density', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.density(300).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert -density 300 "src.jpg" "dst.jpg"');
		});
	});

	describe('#extent', () => {
		it('should be ignored if width or height is not set', () => {
			const cmds = new ImageMagickCommands();
			const widthCmd = cmds.extent({ width: 100 }).get('src.jpg', 'dst.jpg');
			const heightCmd = cmds.extent({ height: 100 }).get('src.jpg', 'dst.jpg');
			widthCmd.should.equal('convert "src.jpg" "dst.jpg"');
			heightCmd.should.equal('convert "src.jpg" "dst.jpg"');
		});

		it('should be applied if both width and height are present', () => {
			const cmds = new ImageMagickCommands();
			const cmd = cmds.extent({ width: 100, height: 200 }).get('src.jpg', 'dst.jpg');
			cmd.should.equal('convert "src.jpg" -extent 100x200 "dst.jpg"');
		});
	});

	describe('#exec', () => {
		beforeEach(() => {
			try {
				fs.unlinkSync('test.jpg'); // eslint-disable-line no-sync
			} catch (e) {}
		});

		it('should run commands', () => {
			const cmds = new ImageMagickCommands();
			return cmds
				.strip()
				.exec(imageFile, 'test.jpg')
				.then(() => {
					isJPG('test.jpg').should.be.true();
				});
		});

		it('should support outputFormat', async () => {
			const cmds = new ImageMagickCommands();
			await cmds.strip().exec(imageFile, 'test.jpg', { format: 'png' });
			isPNG('test.jpg').should.be.true();
		});

		it('should support outputFormat using streams', async () => {
			const cmds = new ImageMagickCommands();
			await cmds.exec(fs.createReadStream(imageFile), fs.createWriteStream('test.jpg', { encoding: 'binary' }), { format: 'png' });
			isPNG('test.jpg').should.be.true();
		});

		it('should support streams', () => {
			const cmds = new ImageMagickCommands();
			return cmds
				.strip()
				.exec(fs.createReadStream(imageFile), fs.createWriteStream('test.jpg', { encoding: 'binary' }))
				.then(() => {
					isJPG('test.jpg').should.be.true();
				});
		});

	});

	describe('#get', () => {
		it('should replace path with - for src and if streams are used', () => {
			const cmds = new ImageMagickCommands();
			const dst = new Writeable();
			const src = new Readable();
			cmds.get(src, dst).should.equal('convert - -');
		});
	});
});


test/imagickal.test.js
'use strict';

const Promise = require('bluebird');
const fs = require('fs');
const ImagickCommands = require('../lib/commands');
const im = require(__dirname + '/../');
const isJPG = require('./helper').isJPG;

require('should');

const imageFile = __dirname + '/fixtures/small.jpg';
const animImage = __dirname + '/fixtures/anim.gif';
const svgFile = __dirname + '/fixtures/small.svg';

describe('Imagick', () => {
	describe('#setDefaults', () => {
		afterEach(() => {
			im.setDefaults({});
		});

		it('should use global defaults', () => {
			im.setDefaults({ executable: 'foobar' });
			im.commands().get('a', 'b').should.equal('foobar "a" "b"');
		});

		it('should ignore global defaults if options is passed to commands()', () => {
			im.setDefaults({ executable: 'foobar' });
			im.commands({ executable: 'monkey' }).get('a', 'b').should.equal('monkey "a" "b"');
		});
	});

	describe('#dimensions', () => {
		it('should support streams', () => {
			return im
				.dimensions(fs.createReadStream(imageFile))
				.then((dim) => {
					dim.width.should.equal(13);
					dim.height.should.equal(10);
					dim.images.should.equal(1);
				});
		});

		it('should return dimensions', () => {
			return im
				.dimensions(imageFile)
				.then((dim) => {
					dim.width.should.equal(13);
					dim.height.should.equal(10);
					dim.images.should.equal(1);
				});
		});

		it('should accept callback function', (done) => {
			im.dimensions(animImage, (err, dim) => {
				if (err) {
					return done(err);
				}

				dim.width.should.equal(64);
				dim.height.should.equal(64);
				dim.images.should.equal(2);
				done();
			});
		});
	});

	describe('#identify', () => {
		it('should support streams', () => {
			return im
				.identify(fs.createReadStream(imageFile))
				.then((data) => {
					data.should.eql({ format: 'jpg', width: 13, height: 10, images: 1 });
				});
		});

		it('should return data in as an object', () => {
			return im
				.identify(imageFile)
				.then((data) => {
					data.should.eql({ format: 'jpg', width: 13, height: 10, images: 1 });
				});
		});

		it('should support extra format options', () => {
			return im
				.identify(imageFile, { format: { orient: '%[orientation]' } })
				.then((data) => {
					data.should.eql({ format: 'jpg', width: 13, height: 10, images: 1, orient: 'TopLeft' });
				});
		});

		it('should accept callback function', (done) => {
			im.identify(animImage, (err, data) => {
				if (err) {
					return done(err);
				}
				data.should.eql({ format: 'gif', width: 64, height: 64, images: 2 });
				done();
			});
		});
	});

	describe('#transform', () => {
		describe('callback function', () => {
			beforeEach(() => {
				this.get = ImagickCommands.prototype.get;
				ImagickCommands.prototype.get = function () {
					return 'echo';
				};
			});

			afterEach(() => {
				ImagickCommands.prototype.get = this.get;
			});


			it('should accept callback function', (done) => {
				im.transform(imageFile, 'dst.jpg', { strip: true }, (err, dst) => {
					if (err) {
						return done(err);
					}

					dst.should.equal('dst.jpg');
					done();
				});
			});
		});

		describe('svg', () => {
			it('should convert svg to jpg', async () => {
				const dst = await im.transform(svgFile, 'dst.jpg', {});
				dst.should.equal('dst.jpg');
				const dimensions = await im.identify(dst);
				dimensions.should.match({
					format: 'jpg',
					width: 534,
					height: 544
				});
			});

			it('should convert svg to png', async () => {
				const dst = await im.transform(svgFile, 'dst.png', {});
				dst.should.equal('dst.png');
				const dimensions = await im.identify(dst);
				dimensions.should.match({
					format: 'png',
					width: 534,
					height: 544
				});
			});

			it('should support variable density', async () => {
				const small = await im.transform(svgFile, 'dst.png', { density: 50 });
				const smallDims = await im.identify(small);

				const large = await im.transform(svgFile, 'dst.png', { density: 150 });
				const largeDims = await im.identify(large);

				smallDims.height.should.be.lessThan(largeDims.height);
				smallDims.width.should.be.lessThan(largeDims.width);
			});
		});

		describe('#transform', () => {
			beforeEach(() => {
				try {
					fs.unlinkSync('test.jpg'); // eslint-disable-line no-sync
				} catch (e) {}
			});

			it('should support streams', () => {
				return im
					.transform(fs.createReadStream(imageFile), fs.createWriteStream('test.jpg', { encoding: 'binary' }), {
						quality: 10,
						strip: true,
						sharpen: { mode: 'variable' }
					})
					.then(() => {
						isJPG('test.jpg').should.be.true();
					});
			});
		});

		describe('command order', () => {
			beforeEach(() => {
				this.exec = ImagickCommands.prototype.exec;
				ImagickCommands.prototype.exec = function () {
					return Promise.resolve(this.commands);
				};
			});

			afterEach(() => {
				ImagickCommands.prototype.exec = this.exec;
			});

			it('should ignore invalid actions', () => {
				return im
					.transform('src.jpg', 'dst.jpg', {
						quality: 10,
						exec: 'no',
						foobar: 'monkey'
					})
					.then((commands) => {
						commands.should.eql(['-quality 10']);
					});
			});

			it('should create commands in order on transform', () => {
				const expected = [
					'-quality 10',
					'-strip',
					'-unsharp 0.8x0.8+1.2+0.05',
					'-filter Catrom -resize 100x\\!',
					'-crop 10x12+1+2'
				];

				return im
					.transform(imageFile, 'dst.jpg', {
						quality: 10,
						strip: true,
						sharpen: { mode: 'variable' },
						resize: { width: 100, flag: '!' },
						crop: { width: 10, height: 12, x: 1, y: 2 },
						rotate: { angle: 20 }
					})
					.then((commands) => {
						commands.should.eql(expected);
					});
			});
		});
	});
});


lib/imagickal.js
'use strict';
const Promise = require('bluebird');
const exec = require('child_process').exec;
const printf = require('util').format;
const stream = require('stream');
const Readable = stream.Readable;
const ImageMagickCommands = require('./commands');
const debug = require('debug')('imagickal');

let globalOpts = {};

const identifyFormatDefaults = {
	format: '%m',
	width: {
		type: Number,
		value: '%w'
	},
	height: {
		type: Number,
		value: '%h'
	}
};

/**
 * Parse output from identify returning record for the first image in sequence
 * The parsed data will contain an extra attribute "images" indicating number of images.
 * @param  {String} output
 * @throws {Error} If unable to parse output
 * @return {Object}
 */
function parseOutput(output) {
	let matches;
	let data;
	let json;
	let images = 0;
	const regx = /(\{[^\}]+\})/g;

	while ((matches = regx.exec(output)) !== null) {
		if (!data) {
			data = matches[1];
		}
		images++;
	}

	if (!data) {
		throw new Error();
	}

	try {
		json = JSON.parse(data);
		if (typeof (json) !== 'object') {
			throw new Error();
		}
	} catch (e) {
		throw new Error();
	}

	json.images = images;
	return json;
}

/**
 * Get Dimensions from an image file.
 * Returns only dimensions for the first image if the image file contains a sequence of images.
 * @param  {String|Readable}   path full path to the file
 * @param {Function} [callback] results in format {width:,height:,images:}
 *                     images attriube shows number of images in sequence. eg for an animated gif
 * @return {Promise} resolved with {width:,height:}
 *                     images attriube shows number of images in sequence. eg for an animated gif
 */
exports.dimensions = function (path, callback) {
	const cmd = 'identify -format "{\\"width\\":%w,\\"height\\":%h}" ' + (path instanceof Readable ? '-' : path);

	const promise = new Promise((resolve, reject) => {
		debug('dimensions', cmd);

		const shell = exec(cmd, Object.assign({ stdio: ['pipe'] }, globalOpts.execOptions || {}), (err, stdout) => {
			if (err) {
				return reject(err);
			}

			try {
				resolve(parseOutput(stdout.toString()));
			} catch (e) {
				return reject(new Error('Unable to parse dimensions from output: ' + stdout.toString()));
			}
		});

		if (path instanceof Readable) {
			path.pipe(shell.stdin);
		}
	});

	return promise.asCallback(callback);
};

/**
 * Identify image format & dimensions on a image.
 * This will also verify that the image file is not corrupt.
 * This will only be done on the first image if the image file contains a sequence of images.
 * @param  {String|Readable} path
 * @param  {Object|Function} options
 * @param  {Object} [options.format] extra format parameters
 * @param  {Boolean} [options.verifyImage] use verbose in identify command to check image for corruption.
 *                                          Callback can be passed as this parameter.
 * @param {Function} [callback] result in format {format:, width:, height:, images:}
 *                     images attribute shows number of images in sequence. eg for an animated gif
 * @return {Promise} - resolved with {format:, width:, height:, images:}
 *                     images attribute shows number of images in sequence. eg for an animated gif
 */
exports.identify = function (path, options = {}, callback) {
	if (typeof (options) === 'function') {
		callback = options;
		options = {};
	}

	const cmd = printf(
		'identify -format "%s" %s%s',
		stringifyIdentifyFormat(Object.assign({}, identifyFormatDefaults, options.format)),
		options.verifyImage ? '-verbose ' : '',
		path instanceof Readable ? '-' : path
	);

	debug('identify', cmd);

	const promise = new Promise((resolve, reject) => {
		const shell = exec(cmd, Object.assign({ stdio: ['pipe'] }, globalOpts.execOptions || {}), (err, stdout) => {
			if (err) {
				if (err.message.match(/decode delegate/)) {
					return reject(new Error('Invalid image file'));
				}
				return reject(err);
			}

			let data;
			try {
				data = parseOutput(stdout.toString());
				data.format = data.format.toLowerCase();
				data.format = data.format === 'jpeg' ? 'jpg' : data.format;
			} catch (e) {
				return reject(new Error('Unable to parse identify data, output was:' + stdout.toString()));
			}

			if (!data) {
				return reject(new Error('Unable to identify image, output was:' + stdout.toString()));
			}

			resolve(data);
		});

		if (path instanceof Readable) {
			path.pipe(shell.stdin);
		}
	});

	return promise.asCallback(callback);
};

function stringifyIdentifyFormat(format) {
	const strings = Object.keys(format).reduce((acc, key) => {
		const val = format[key];
		const strValue = typeof (val) === 'object' && val.type === Number ? val.value : `\\"${val}\\"`;
		acc.push(`\\"${key}\\":${strValue}`);
		return acc;
	}, []);

	return `{${strings.join(',')}}`;
}

/**
 * Create new ImageMagickCommands instances
 * @param {Object} [opts] if specified it not use any of the global options set by setDefaults()
 * @param {String} [opts.executable] path/command for convert.
 * @return {ImageMagickCommands}
 */
const commands = exports.commands = function (opts) {
	return new ImageMagickCommands(opts || globalOpts);
};


/**
 * TRANSFORMATION
 */
const invalidActions = ['get', 'exec'];

/**
 * Create ImageMagickCommands object and apply actions on it.
 * @param  {Object} actions
 * @return {ImageMagickCommands}
 */
function applyActions(actions) {
	const cmds = commands();
	Object.keys(actions).forEach((action) => {
		if (invalidActions.indexOf(action) === -1 && typeof (cmds[action]) === 'function') {
			cmds[action](actions[action]);
		}
	});
	return cmds;
}

/**
 * Calculate new dimensions on image after transformation based on its actions.
 * @param  {Object} actions
 * @return {Object|Boolean} {width:, height:}, false if no new dimensions could be calculated
 */
function calculateNewDimensions(actions) {
	if (actions.resize && actions.resize.width && actions.resize.height) {
		return { width: actions.resize.width, height: actions.resize.height };
	}

	if (actions.resize && actions.resize.width) {
		return {
			width: actions.resize.width
		};
	}

	if (actions.resize && actions.resize.height) {
		return {
			height: actions.height
		};
	}

	return false;
}

/**
 * Transform image
 * @param  {String} src
 * @param  {String} dst
 * @param  {Object} actions
 * @param {Object} [opts]
 * @param {Number} [opts.format] output format jpg, png
 * @param  {Function} [callback]
 * @return {Promise} resolved promise with dst as value
 */
exports.transform = function (src, dst, actions, opts, callback) {
	if (actions.sharpen && actions.sharpen.mode === 'variable') {
		const newDim = calculateNewDimensions(actions);
		if (newDim) {
			const moddedActions = JSON.parse(JSON.stringify(actions));
			moddedActions.sharpen.width = newDim.width;
			moddedActions.sharpen.height = newDim.height;

			return applyActions(moddedActions).exec(src, dst, opts, callback);
		}
	}

	return applyActions(actions).exec(src, dst, opts, callback);
};

/**
 * Set default options to be passed to ImageMagickCommands.
 * @param {Object} opts
 */
exports.setDefaults = function (opts) {
	globalOpts = opts;
};


lib/commands.js
'use strict';
const Promise = require('bluebird');
const exec = require('child_process').exec;
const	printf = require('util').format;
const debug = require('debug')('imagickal');

/**
 * Differet sharpening presets, values are (radius, amount, threshold)
 * @type {Object}
 */
const sharpening = {
	light: [0.5, 1, 0.05],
	moderate: [0.65, 1.1, 0.05],
	strong: [0.8, 1.2, 0.05],
	extreme: [1.0, 1.5, 0.00]
};

/**
 * Valid resize flags, http://www.imagemagick.org/Usage/resize/#noaspect
 * @type {Array}
 */
const resizeFlags = ['<', '>', '!', '^'];

/**
 * valid gravity values, http://www.imagemagick.org/script/command-line-options.php#gravity
 * @type {Array}
 */
const gravities = ['NorthWest', 'North', 'NorthEast', 'West', 'Center', 'East', 'SouthWest', 'South', 'SouthEast'];

/**
 * Check if variable is a number
 * @param  {Mixed}  n [description]
 * @return {Boolean}
 */
function isNumber(n) {
	return !isNaN(parseFloat(n)) && isFinite(n);
}

/**
 * Calculate what sharpening mode to use base on image dimensions
 * @param  {Integer} width
 * @param  {Integer} height
 * @return {String} the sharpening mode
 */
function getVariableMode(width, height) {
	if ((width && width <= 50) || (height && height <= 50)) {
		return 'extreme';
	}

	if ((width && width <= 100) || (height && height <= 100)) {
		return 'strong';
	}

	if ((width && width <= 300) || (height && height <= 300)) {
		return 'moderate';
	}

	if ((width && width <= 500) || (height && height <= 500)) {
		return 'light';
	}
	return false;
}

/**
 * Sanitize Resize flag option, quote if needed.
 * @param  {String} flag
 * @return {String} empty string if invalid flag
 */
function getResizeFlag(flag) {
	if (resizeFlags.indexOf(flag) < 0) {
		return '';
	}

	return flag === '^' ? flag : '\\' + flag;
}

function pipeOrPath(obj, format) {
	format = format ? `${format}:` : '';
	const quoted = typeof obj === 'string' ? `"${obj}"` : obj;
	return typeof (obj.pipe) === 'function' ? `${format}-` : `${format}${quoted}`;
}

function ImageMagickCommands(opts) {
	opts = opts || {};
	this.executable = opts.executable || 'convert';
	this.inputOptions = [];
	this.commands = [];
	this.maxBuffer = opts.maxBuffer || 1024 * 1024 * 100;
	this.execOptions = opts.execOptions || {};
}

/**
 * Set gravity
 * See http://www.imagemagick.org/script/command-line-options.php#gravity for valid gravities.
 *
 * @param  {String} gravity
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.gravity = function (gravity) {
	if (gravities.indexOf(gravity) < 0) {
		return this;
	}

	this.commands.push('-gravity ' + gravity);
	return this;
};

/**
 * Resize, if only one param is supplied it will remain aspect ratio.
 * @param  {Object} opts {width, height, flag, gravity}
 * @param {String} opts.flag imagemagick resize flags (<,>,!,^)
 * @param {String} opts.gravity imagemagick gravity (NorthWest, North, NorthEast, West, Center, East, SouthWest, South, SouthEast)
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.resize = function (opts) {
	if (!opts.width && !opts.height) {
		return this;
	}
	const height = opts.height || '';
	const width = opts.width || '';
	const flag = opts.flag ? getResizeFlag(opts.flag) : '';

	this.commands.push(printf('-filter Catrom -resize %sx%s%s', width, height, flag));
	return this;
};

/**
 * Crop
 * @param  {Object} opts {width,height,x,y}
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.crop = function (opts) {
	if (!opts.width || !opts.height || !isNumber(opts.x) || !isNumber(opts.y)) {
		return this;
	}
	const x = opts.x >= 0 ? '+' + opts.x : opts.x;
	const y = opts.y >= 0 ? '+' + opts.y : opts.y;

	this.commands.push(printf('-crop %sx%s%s%s', opts.width, opts.height, x, y));
	return this;
};

/**
 * Extent
 * @param  {Object} opts {width, height}
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.extent = function (opts) {
	if (!opts.width || !opts.height) {
		return this;
	}
	this.commands.push(printf('-extent %sx%s', opts.width, opts.height));
	return this;
};

/**
 * Rotate
 * @param  {Object} opts {angle,x,y,bgColor?}
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.rotate = function (opts) {
	if (!opts.angle || !opts.x || !opts.y) {
		return this;
	}

	const bgColor = opts.bgColor ? '-background ' + opts.bgColor + ' -virtual-pixel background ' : '';
	this.commands.push(printf(
		'%s-distort ScaleRotateTranslate \'%s,%s %s\'',
		bgColor, opts.x, opts.y, opts.angle
	));
	return this;
};

/**
 * Sharpen
 * mode - available presets are [light,moderate,strong,extreme,variable,off]
 * width,height only needed if mode is variable
 * @param  {Object} opts {mode, width?, height?]
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.sharpen = function (opts) {
	let mode = opts.mode;

	if (Object.keys(sharpening).concat(['off', 'variable']).indexOf(mode) < 0) {
		return this;
	}

	if (mode === 'variable') {
		mode = getVariableMode(opts.width, opts.height);
	}

	if (mode === 'off' || !mode) {
		return this;
	}

	const radius = sharpening[mode][0];
	const amount = sharpening[mode][1];
	const threshold = sharpening[mode][2];
	const sigma = radius < 1 ? radius : Math.sqrt(radius);

	this.commands.push(printf('-unsharp %dx%d+%d+%d', radius, sigma, amount, threshold));
	return this;
};

/**
 * Quality
 * @param  {Number} quality
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.quality = function (quality) {
	if (!isNumber(quality)) {
		return this;
	}

	this.commands.push('-quality ' + quality);
	return this;
};

/**
 * Density
 * @param  {Number} density
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.density = function (density) {
	if (!isNumber(density)) {
		return this;
	}

	this.inputOptions.push('-density ' + density);
	return this;
};

/**
 * Strip
 * @return {ImageMagickCommands}
 */
ImageMagickCommands.prototype.strip = function () {
	this.commands.push('-strip');
	return this;
};

/**
 * Get the commandline string
 * @param {String} src
 * @param {String} dst
 * @param {String} [outputFormat]
 * @return {String}
 */
ImageMagickCommands.prototype.get = function (src, dst, outputFormat) {
	return [this.executable].concat(this.inputOptions, [pipeOrPath(src)], this.commands, [pipeOrPath(dst, outputFormat)]).join(' ');
};

/**
 * Execute the commands
 * @param  {String|Readable} src
 * @param  {String|Writeable} dst
 * @param {Object} [opts]
 * @param {Number} [opts.format] output format jpg, png
 * @param  {Function} [callback] optional for those that dont want promise.
 * @return {Promise}
 */
ImageMagickCommands.prototype.exec = function (src, dst, opts = {}, callback) {
	if (typeof (opts) === 'function') {
		callback = opts;
		opts = {};
	}

	debug('Executing', this.get(src, dst, opts.format));

	const execOpts = Object.assign(
		{ stdio: ['pipe', 'pipe'], maxBuffer: this.maxBuffer, encoding: 'binary' },
		this.execOptions
	);

	const promise = new Promise((resolve, reject) => {
		const shell = exec(this.get(src, dst, opts.format), execOpts, (err) => {
			if (err) {
				return reject(err);
			}

			resolve(dst);
		});

		if (typeof (src.pipe) === 'function') {
			src.pipe(shell.stdin);
		}

		if (typeof (dst.pipe) === 'function') {
			shell.stdout.pipe(dst);
		}
	});

	return promise.asCallback(callback);
};

module.exports = ImageMagickCommands;
